File: agoda/api/hotels.js
const buildHotelApiUrl = (id, checkIn, checkOut, adult, children, room, filters) => {
    const baseUrl = 'http://localhost:8080/api/hotels';
    const params = new URLSearchParams({
        id,
        checkIn,
        checkOut,
        adult,
        room
    });
    if (children.length > 0) {
        if (children.length === 1) {
            params.append('children', children[0]);
        } else {
            params.append('children', children.join(','));
        }
    }
    if (filters.starRating.length > 0) {
        params.append('starRating', filters.starRating.join(','));
    }
    if (filters.guestRating.length > 0) {
        params.append('guestRating', filters.guestRating.join(','));
    }
    if (filters.propertyFacilities.length > 0) {
        params.append('facilities', filters.propertyFacilities.join(','));
    }
    if (filters.bedrooms.length > 0) {
        params.append('bedrooms', filters.bedrooms.join(','));
    }
    if (filters.paymentOptions.length > 0) {
        params.append('paymentOptions', filters.paymentOptions.join(','));
    }
    if (filters.popularWithFamilies) {
        params.append('kidsStayForFree', '250691');
    }
    return `${baseUrl}?${params.toString()}`;
};

const fetchHotelData = async (id, checkIn, checkOut, adult, children, room, filters) => {
    const url = buildHotelApiUrl(id, checkIn, checkOut, adult, children, room, filters);

    try {
        const response = await fetch(url);

        if (!response.ok) {
            return null;
        }

        const data = await response.json();
        return data;
    } catch (error) {
        console.error('Error fetching hotel data:', error);
        return null;
    }
};

export default fetchHotelData;

File: handlers/hotels.go
package handlers

import (
	"encoding/json"
	"hotel/models"
	"io"
	"log"
	"net/http"

	"github.com/gofiber/fiber/v2"
)

func GetHotels(c *fiber.Ctx) error {
	id := c.Query("id")
	checkIn := c.Query("checkIn")
	checkOut := c.Query("checkOut")
	adults := c.Query("adult")
	room := c.Query("room")
	searchUrl := url + "search-overnight?id=" + id + "&checkinDate=" + checkIn + "&checkoutDate=" + checkOut + "&adult=" + adults
	searchUrl += "&room=" + room
	rating := c.Query("starRating")
	childrens := c.Query("children")
	guestReviews := c.Query("guestReviews")
	kidsStayForFree := c.Query("kidsStayForFree")
	if rating != "" {
		searchUrl += "&starRating=" + rating
	}
	paymentOptions := c.Query("paymentOptions")
	if paymentOptions != "" {
		searchUrl += "&paymentOptions=" + paymentOptions
	}
	facilities := c.Query("facilities")
	if facilities != "" {
		searchUrl += "&facilities=" + facilities
	}
	bedrooms := c.Query("bedrooms")
	if bedrooms != "" {
		searchUrl += "&bedrooms=" + bedrooms
	}
	if guestReviews != "" {
		searchUrl += "&guestReviews=" + guestReviews
	}
	if kidsStayForFree != "" {
		searchUrl += "&kidsStayForFree=" + kidsStayForFree
	}
	if childrens != "" {
		searchUrl += "&childAges=" + childrens
	}
	req, err := http.NewRequest(http.MethodGet, searchUrl, nil)
	if err != nil {
		log.Println(err)
		return c.SendString("Internal server error")
	}
	req.Header.Set("x-rapidapi-key", key)
	log.Println(searchUrl)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		log.Println(err)
		return c.Status(fiber.StatusInternalServerError).SendString("Internal server error")
	}
	if resp.StatusCode != http.StatusOK {
		log.Println(resp.Status)
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			log.Println(err)
			return c.Status(fiber.StatusInternalServerError).SendString("Internal server error")
		}
		log.Println(string(body))
		return c.Status(fiber.StatusInternalServerError).SendString("Internal server error")
	}
	var hotelResponse models.ResponseHotels
	err = json.NewDecoder(resp.Body).Decode(&hotelResponse)
	if err != nil {
		log.Println(err)
		return c.Status(fiber.StatusInternalServerError).SendString("Internal server error")
	}
	var hotelList []models.HotelListing
	for _, property := range hotelResponse.Data.Properties {
		price := 0.0
		currency := "USD"
		if len(property.Pricing.Offers) > 0 {
			price = property.Pricing.Offers[0].RoomOffers[0].Room.Pricing[0].Price.PerNight.Inclusive.Display
			currency = property.Pricing.Offers[0].RoomOffers[0].Room.Pricing[0].Currency
		}
		hotelList = append(hotelList, models.GetHotelListing(property, price, hotelResponse.Data.Filters, currency))
	}
	return c.JSON(hotelList)
}
